src/agents/user_agent.py

```diff
@@ -1,12 +1,126 @@
+        template.set_metadata("performative", "inform")
+        template.set_metadata("action", "consolidated-offers")
+        return template
+    
+    def _prepare_reservation_response_template(self):
+        template = Template()
+        template.to = f"{self.jid}@{DEFAULT_HOST}"
+        template.set_metadata("performative", "inform")
+        template.set_metadata("action", "reservation-response")
+        return template
+
+    def add_coordinator(self, coordinator_jid):
+        self.coordinators_jids.append(coordinator_jid)
+
+    async def request_parking_offers(self, x, y, time_start, time_stop):
+
+        self._validate_location(x, y)
+        self._validate_timeslot(time_start, time_stop)
+
+        self.pending_reservation = Reservation(x, y, time_start, time_stop)
+
+        check_offers_msg = CheckOffers(
+            x=x,
+            y=y,
+            time_start=time_start,
+            time_stop=time_stop
+        )
+        for coordinator_jid in self.coordinators_jids:
+            to_send = Message(
+                to=coordinator_jid,
+                body=json.dumps(check_offers_msg.dict()),
+                metadata={"performative": "query-ref", "action": "check-offers"}
+            )
+            await self.send(to_send)
+
+    def choose_parking_offer(self, offers: List[dict]):
+        #TODO implement logic of choosing an offer
+        return offers[0]["parking_id"] if offers else None
+
+    class MakeReservation(CyclicBehaviour):
+        """Behaviour for making a reservation
+            After receiving consolidated offers, the user chooses one offer and makes a reservation.
+        """
+        async def run(self):
+            msg = await self.receive(timeout=RECEIVE_TIMEOUT)
+            if msg:
+                consolidated_offers = ConsolidatedOffers(**json.loads(msg.body))
+                chosen_parking_id = self.agent.choose_parking_offer(consolidated_offers.offers)
+                if self.agent.pending_reservation:
+                    self.agent.pending_reservation.parking_id = chosen_parking_id
+                    res = self.agent.pending_reservation
+                    reservation_request = RequestReservation(res.time_start, res.time_stop, chosen_parking_id, self.agent.jid)
+                    to_send = Message(to=consolidated_offers.sender, body=json.dumps(reservation_request.dict()), metadata={"performative": "request", "action": "make-reservation"})
+                    await self.send(to_send)
+
+    class AwaitReservationConfirmation(CyclicBehaviour):
+        """Behaviour for waiting for reservation confirmation"""
         async def run(self):
-            # @TODO to be implemented
-            pass
+            msg = await self.receive(timeout=RECEIVE_TIMEOUT)
+            if msg:
+                reservation_response = ReservationResponse(**json.loads(msg.body))
+                if reservation_response.success:
+                    #TODO What about reservation id?
```

Comment:
How about a timestamp + random number for reservation id? Or simply a random number.

Reply:
I guess the reservation ids should be unique across all users. I was thinking about making a hash using the user's id and maybe timestamp?